### PR TITLE
Change UploadDataInterface to use ReadIntoBuffer

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
Instead of using `GetUploadDataBuffer` and `RefreshUploadDataBuffer`, we
changed the `UploadDataInterface` to use the `ReadIntoBuffer` method
which reads at most `max_bytes` data into a provided buffer.

http://b/204282170

### CODE REVIEW

File: src/ProducerEventProcessor/UploaderClientCaptureEventCollector.cpp [hunk 1]
```diff
@@ -65,49 +65,67 @@ void UploaderClientCaptureEventCollector::AddEvent(ClientCaptureEvent&& event) {
...
+  return DataReadiness::kEndOfData;
+}
+
+size_t UploaderClientCaptureEventCollector::ReadIntoBuffer(void* dest, size_t max_bytes) {
+  absl::MutexLock lock(&mutex_);
+
+  size_t bytes_to_read = std::min(capture_data_to_upload_.size() - byte_position_, max_bytes);
+
+  char* dest_buffer = static_cast<char*>(dest);
+  memcpy(dest_buffer, capture_data_to_upload_.data() + byte_position_, bytes_to_read);
```
Comment (beckerhe) [inline src/ProducerEventProcessor/UploaderClientCaptureEventCollector.cpp:125]: Nit: `std::memcpy`

File: src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp [hunk 2]
```diff
@@ -82,14 +82,13 @@ class UploaderClientCaptureEventCollectorTest : public testing::Test {
     produce_finished_ = true;
   }
 
+  static constexpr size_t kUploadBufferSize = 100;
```
Comment (beckerhe) [inline src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp:85]: Nit: `constexpr` implies `const` and `const` implies `static`. So technically the `static` is not needed.

File: src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp [hunk 3]
```diff
@@ -98,6 +97,7 @@ class UploaderClientCaptureEventCollectorTest : public testing::Test {
 
   std::thread data_upload_thread_;
   std::thread data_produce_thread_;
+  char* upload_buffer_ = new char[kUploadBufferSize];
```
Comment (beckerhe) [inline src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp:100]: What about `std::array<char, kUploadBufferSize> upload_buffer_` instead?

**Review by beckerhe — APPROVED** [review]
Awesome. Otherwise LGTM

File: src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp (hunk 2 above)
Comment (vickyliu-go4it) [reply src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp:85]: Thanks for the suggestion! 

But due to "non-static data member cannot be constexpr", I might keep `static` here.

Comment (beckerhe) [reply src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp:85]: I think you're confusing the different usages of `static`. `static` means something different depending on where it appears: in a class declaration, in a function body, on the global level. What you're citing only refers to "non-static data members", meaning when `static` is used in a class declaration.

Comment (beckerhe) [reply src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp:85]: I just realized that this is in an anonymoys class. Never mind. Sorry. :see_no_evil:

Comment (vickyliu-go4it) [reply src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp]: Update: Moved `kUploadBufferSize` outside of the class and removed `static` for this and other constants.

**Pull request merged**